File: ai-chat/llm/provider/openai/openai.go
// OpenAI is an llm provider implementation for the OpenAI API.
package openai

import (
	"context"
	"encoding/base64"
	"io"
	"sync"

	"github.com/cockroachdb/errors"
	"github.com/sashabaranov/go-openai"

	"encore.app/llm/provider"
	"encore.app/pkg/fns"
	"encore.dev/config"
	"encore.dev/pubsub"
	"encore.dev/rlog"
	"encore.dev/types/uuid"
)

var TopicRef = pubsub.TopicRef[pubsub.Publisher[*provider.BotResponse]](provider.LLMMessageTopic)

// This uses Encore's built-in secrets manager, learn more: https://encore.dev/docs/primitives/secrets
var secrets struct {
	OpenAIKey string
}

type Config struct {
	ChatModel   config.String
	ImageModel  config.String
	MaxTokens   config.Int
	Temperature config.Float32
	TopP        config.Float32
}

// This uses Encore Configuration, learn more: https://encore.dev/docs/develop/config
var cfg = config.Load[*Config]()

// This declares a Encore Service, learn more: https://encore.dev/docs/primitives/services-and-apis/service-structs
//
//encore:service
type Service struct {
	client      *openai.Client
	taskCtx     context.Context
	mu          sync.Mutex
	activeTasks map[string]*task
}

// initService initializes the OpenAI service by creating a client.
func initService() (*Service, error) {
	if secrets.OpenAIKey == "" {
		return nil, nil
	}
	svc := &Service{
		taskCtx:     context.Background(),
		activeTasks: map[string]*task{},
		client:      openai.NewClient(secrets.OpenAIKey),
	}
	return svc, nil
}

// Ping returns an error if the service is not available.
// encore:api private
func (p *Service) Ping(ctx context.Context) error {
	if p == nil {
		return errors.New("OpenAI service is not available. Add OpenAIKey secret to enable it.")
	}
	return nil
}

type GenerateAvatarRequest struct {
	Prompt string
}

type GenerateAvatarResponse struct {
	Image []byte
}

// GenerateAvatar generates an avatar image based on the given prompt. The model is configurable in the config.
//
//encore:api private method=POST path=/openai/generate-avatar
func (p *Service) GenerateAvatar(ctx context.Context, req *GenerateAvatarRequest) (*GenerateAvatarResponse, error) {
	resp, err := p.client.CreateImage(ctx, openai.ImageRequest{
		Prompt:         req.Prompt,
		Model:          cfg.ImageModel(),
		N:              1,
		Quality:        "standard",
		Size:           openai.CreateImageSize1024x1024,
		ResponseFormat: openai.CreateImageResponseFormatB64JSON,
	})
	if err != nil {
		return nil, errors.Wrap(err, "create image")
	}
	data, err := base64.StdEncoding.DecodeString(resp.Data[0].B64JSON)
	if err != nil {
		return nil, errors.Wrap(err, "decode image")
	}
	return &GenerateAvatarResponse{Image: data}, nil
}

type AskRequest struct {
	Message string
}

type AskResponse struct {
	Message string
}

// Ask sends a single message to the OpenAI chat model and returns the response.
//
//encore:api private method=POST path=/openai/ask
func (p *Service) Ask(ctx context.Context, req *AskRequest) (*AskResponse, error) {
	modResp, err := p.client.Moderations(ctx, openai.ModerationRequest{
		Input: req.Message,
	})
	if err != nil {
		return nil, errors.Wrap(err, "moderations")
	}
	if fns.Any(modResp.Results, func(r openai.Result) bool { return r.Flagged }) {
		return nil, errors.New("message was flagged by OpenAI")
	}
	resp, err := p.client.CreateChatCompletion(ctx, openai.ChatCompletionRequest{
		Model: cfg.ChatModel(),
		Messages: []openai.ChatCompletionMessage{
			{
				Role:    openai.ChatMessageRoleUser,
				Content: req.Message,
			},
		},
		MaxTokens:   cfg.MaxTokens(),
		N:           1,
		Temperature: cfg.Temperature(),
		TopP:        cfg.TopP(),
	})
	if err != nil {
		return nil, err
	}
	return &AskResponse{Message: resp.Choices[0].Message.Content}, nil
}

// CancelTask cancels an active task by ID. It's used to cancel queued chat completions.
//
//encore:api private method=DELETE path=/openai/task/:taskID
func (p *Service) CancelTask(ctx context.Context, taskID string) error {
	p.mu.Lock()
	defer p.mu.Unlock()
	if task, ok := p.activeTasks[taskID]; ok {
		task.cancel()
		delete(p.activeTasks, taskID)
	}
	return nil
}

// ContinueChat continues a chat conversation with the OpenAI chat model. The responses are streamed back to the chat service
// using the a pubsub topic
//
//encore:api private method=POST path=/openai/continue-chat
func (p *Service) ContinueChat(ctx context.Context, req *provider.ChatRequest) (*provider.ContinueChatResponse, error) {
	var messages []openai.ChatCompletionMessage
	if req.SystemMsg != "" {
		messages = append(messages, openai.ChatCompletionMessage{
			Role:    openai.ChatMessageRoleSystem,
			Content: req.SystemMsg,
		})
	}
	for _, m := range req.Messages {
		role := openai.ChatMessageRoleUser
		if req.FromBot(m) {
			// The bot is always the assistant in this case and should reply without prefixes
			role = openai.ChatMessageRoleAssistant
		}

		messages = append(messages, openai.ChatCompletionMessage{
			Role:    role,
			Content: req.Format(m),
		})
	}

	task := &task{id: uuid.Must(uuid.NewV4()).String(), service: p}
	p.mu.Lock()
	p.activeTasks[task.id] = task
	p.mu.Unlock()
	go task.run(p.taskCtx, p.client, messages, req.Write)
	return &provider.ContinueChatResponse{
		TaskID: task.id,
	}, nil
}

type task struct {
	id      string
	service *Service
	cancel  func()
}

func (t *task) run(ctx context.Context, client *openai.Client, messages []openai.ChatCompletionMessage, writer func(context.Context, string) error) {
	ctx, t.cancel = context.WithCancel(ctx)
	stream, err := client.CreateChatCompletionStream(ctx, openai.ChatCompletionRequest{
		Model:     cfg.ChatModel(),
		Messages:  messages,
		MaxTokens: cfg.MaxTokens(),
		N:         1,
	})
	if err != nil {
		rlog.Error("create chat completion stream", "error", err)
		return
	}
	defer fns.CloseIgnore(stream)
	for {
		response, err := stream.Recv()
		select {
		case <-ctx.Done():
			return
		default:
		}
		if errors.Is(err, io.EOF) {
			err := writer(ctx, "\n")
			if err != nil {
				rlog.Warn("write response", "error", err)
			}
			break
		}
		err = writer(ctx, response.Choices[0].Delta.Content)
		if err != nil {
			rlog.Warn("write response", "error", err)
			return
		}
	}
	t.service.mu.Lock()
	defer t.service.mu.Unlock()
	delete(t.service.activeTasks, t.id)
}

File: ts/ai-chat/llm/provider/openai/openai.ts
import OpenAI from "openai";
import { secret } from "encore.dev/config";
import { api } from "encore.dev/api";
import { LLMMessage, Meta, ResponseBuilder } from "../provider";
import { botEventTopic } from "../../service/topics";

const model = "gpt-4o";
const openAIKey = secret("OpenAIKey");
const client = new OpenAI({ apiKey: openAIKey() });

// available checks if the OpenAI provider is configured
export const available = api(
  { expose: false, path: "/openai/available", method: "GET" },
  async function (): Promise<{ available: boolean }> {
    return { available: !!openAIKey() };
  },
);

// chat is the main function to interact with the OpenAI chat model
export const chat = api(
  { expose: false, path: "/openai/chat", method: "POST" },
  async function ({
    system,
    messages,
    meta,
  }: {
    system: string;
    messages: LLMMessage[];
    meta: Meta;
  }): Promise<void> {
    let allMessages: [
      { role: "system" | "user" | "assistant"; content: string },
    ] = [{ role: "system", content: system }];
    allMessages.push(...messages);
    const stream = await client.chat.completions.create({
      messages: allMessages,
      model: model,
      n: 1,
      stream: true,
    });
    const builder = new ResponseBuilder(meta, (msg) => {
      return botEventTopic.publish(msg);
    });
    for await (const completion of stream) {
      const content = completion.choices[0]?.delta?.content || "";
      await builder.parse(content);
    }
    await builder.parse("\n");
  },
);

// ask is used for one-off questions to the OpenAI model
export const ask = api(
  { expose: false, path: "/openai/ask", method: "POST" },
  async function ({
    prompt,
  }: {
    prompt: string;
  }): Promise<{ response: string }> {
    const chatCompletion = await client.chat.completions.create({
      messages: [{ role: "user", content: prompt }],
      model: model,
      n: 1,
    });
    const response = chatCompletion.choices.at(0)?.message.content;
    if (!response) {
      return Promise.reject("no response");
    }
    return { response };
  },
);

// generateAvatar is used to generate an avatar image from a prompt
export const generateAvatar = api(
  { expose: false, path: "/openai/image", method: "POST" },
  async function ({ prompt }: { prompt: string }): Promise<{ data: string }> {
    const imageResponse = await client.images.generate({
      prompt: prompt,
      size: "1024x1024",
      model: "dall-e-3",
      n: 1,
      response_format: "b64_json",
    });
    const data = imageResponse.data.at(0)?.b64_json;
    if (!data) {
      return Promise.reject("no image returned");
    }
    return { data };
  },
);
